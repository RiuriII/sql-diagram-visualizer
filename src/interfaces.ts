export interface Column {
    name: string;
    type: string;
}

export interface ForeignKey {
    foreignKey: string;
    referenceTable: string;
}

export interface Table {
    tableName: string;
    column: Column[];
    foreignKey: ForeignKey[];
}

export interface SvgTable {
    tableName: string;
    tableMarkup: string;
    posX: number;
    posY: number;
    tableWidth: number;
    tableHeight: number;
    columns: Column[];
    foreignKey: ForeignKey[];
}

export interface Connection {
    sourcePosX: number;
    sourcePosY: number;
    targetPosX: number;
    targetPosY: number;
    color: string;
}
