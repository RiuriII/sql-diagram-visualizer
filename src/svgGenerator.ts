import * as Handlebars from "handlebars";
import { Column, Connection, SvgTable, Table } from "./interfaces";

// SVG template  
const svgTemplate = `
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 3000 1000" preserveAspectRatio="xMinYMin meet">
{{#each connections}}
      <line x1="{{sourcePosX}}" y1="{{sourcePosY}}" x2="{{targetPosX}}" y2="{{targetPosY}}" stroke="{{color}}" stroke-width="3" stroke-linecap="round" marker-end="url(#arrowhead)"/>
  {{/each}}
  {{#each tables}}
  <g id="table-{{@index}}" transform="translate({{posX}}, {{posY}})">
  {{{tableMarkup}}}
  </g>
  {{/each}}
</svg>
`;

// Table template
const tableTemplate = `
    <rect x="0" y="0" width="{{tableWidth}}" height="{{tableHeight}}" fill="#FFFFFF" stroke="black" filter="url(#shadow)"/>
    <rect x="0" y="0" width="{{tableWidth}}" height="50" fill="#8B29A6" />
    <text x="{{centerX}}" y="30" font-family="Arial, sans-serif" font-size="20" font-variant="small-caps" text-anchor="middle" font-weight="bold" fill="#ffffff">{{tableName}}</text>
    {{#each column}}
      <rect x="0" fill="{{#if (isOdd @index)}}#f5f5f5{{else}}#ffffff{{/if}}" y="{{calculateY @index}}" width="{{reduceLine ../tableWidth}}" height="40" />
      <text x="20" y="{{calculateY @index}}" dy="25" font-family="Arial, sans-serif" font-size="16" fill="black" font-weight="bold">{{name}}     <tspan dx="30" font-weight="normal">{{type}}</tspan></text>
    {{/each}}
    <defs>
  <filter id="shadow" x="-10%" y="-10%" width="120%" height="120%">
    <feDropShadow dx="3" dy="3" stdDeviation="4" flood-color="#000" flood-opacity="0.2" />
  </filter>
</defs>
`;

// Helper to calculate the center of the table
Handlebars.registerHelper('centerX', function (tableWidth: any) {
    return tableWidth.data.root.tableWidth / 2;
});

// Helper to reduce the line width
Handlebars.registerHelper('reduceLine', function (tableWidth: number) {
    return tableWidth - 1;
});

// Helper to calculate the Y position of each column
Handlebars.registerHelper('calculateY', function (index: number) {
    return 50 + index * 40; // Ajuste o espaçamento conforme necessário
});

// Helper to check if the index is odd
Handlebars.registerHelper('isOdd', function (index: number) {
    return index % 2 === 1;
});

// Estimates that each character occupies about 0.6 * fontSize pixels in width
const estimateTextWidth = (text: string, fontSize: number) => text.length * fontSize * 0.6;

// Calculates the width of the table based on the column names and types
const calculateTableWidth = (columns: Column[], fontSize: number) => {
    let maxWidth = 0;
    const paddingWidth = 40;
    columns.forEach(column => {
        const columnText = `${column.name}: ${column.type}`;
        const width = estimateTextWidth(columnText, fontSize);
        if (width > maxWidth) {
            maxWidth = width;
        }
    });
    return Math.ceil(maxWidth + paddingWidth); // Add padding to the width
};

// Generates connections between tables based on foreign keys
const connectionsGenerator = (data: Table[], tables: SvgTable[]): Connection[] => {
    const connections: Connection[] = [];

    // Factor to adjust the Y position of connections
    const heightAdjustmentFactor = 1.5;

    data.forEach((table, index: number) => {
        table.foreignKey.forEach((fk) => {
            const sourceTable = tables[index];   // Table with the foreign key
            const targetTable = tables.find(t => t.tableName === fk.referenceTable); // Table referenced by the foreign key
            if (targetTable) {

                // Origin point in the foreign key table (right side of the table)
                const sourcePosX = sourceTable.posX + sourceTable.tableWidth;
                const sourcePosY = sourceTable.posY + sourceTable.tableHeight;

                // Destination point in the referenced table (left side of the table)
                const targetPosX = targetTable.posX + targetTable.tableHeight / heightAdjustmentFactor;
                const targetPosY = targetTable.posY + targetTable.tableHeight / heightAdjustmentFactor;

                const palette = ['#8B29A6', '#0D0C0C', '#444DF2', '#BA1511', '#0692E1', '#F58804'];

                const color = palette[index % palette.length];  // Choose color from palette based on index

                // Add the connection between the tables
                connections.push({ sourcePosX, sourcePosY, targetPosX, targetPosY, color });
            }
        });
    });

    return connections;
};

// Calculates the positions of tables based on padding and window width
const calculateTablePositions = (padding: {x: number, y: number}, tableWidth: number, tableHeight: number, previousTableWidth: number, windowWidth: number, currentPosition: { x: number, y: number }) => {

    // If `currentPosition` exceeds the allowed total width, move to the next "row" (Y axis)
    if (currentPosition.x + tableWidth + padding.x > windowWidth) {
        currentPosition.x = 100; // Reset X position
        currentPosition.y += tableHeight + padding.y; // Move Y position to next row
    } else {
        currentPosition.x += previousTableWidth + padding.x; // Continue on the same row
    }

    return currentPosition;
};

const tablesGenerator = (tables: Table[]) => {
    const windowWidth = 2500; // Total width of the SVG
    const tableTemplateCompiled = Handlebars.compile(tableTemplate);
    let currentPosition = { x: 100, y: 100 };
    const padding = { x: 100, y: 100 }; // Space between tables

    return tables.map((table, index) => {
        const fontSize = 16; // Font size of the column names
        const rowHeight = 35;
        const tableHeight = padding.y + table.column.length * rowHeight; // Calculate the height of the table
        const tableWidth = calculateTableWidth(table.column, fontSize); // Calculate the width of the table

        if (index > 0) {
            const previousTableWidth = calculateTableWidth(tables[index - 1].column, fontSize);

            currentPosition = calculateTablePositions(padding, tableWidth, tableHeight, previousTableWidth, windowWidth, currentPosition);

        }

        const posX = currentPosition.x;
        const posY = currentPosition.y;

        return {
            tableName: table.tableName,
            tableMarkup: tableTemplateCompiled({ ...table, tableHeight, tableWidth }),
            posX,
            posY,
            tableWidth,
            tableHeight,
            columns: table.column,
            foreignKey: table.foreignKey
        };
    });
};

export const svgGenerator = (databaseTables: Table[]) => {

    if (databaseTables.length === 0) {
        throw new Error('No tables found');
    }

    const svgTemplateCompiled = Handlebars.compile(svgTemplate);

    const tables: SvgTable[] = tablesGenerator(databaseTables);

    const connections: Connection[] = connectionsGenerator(databaseTables, tables);

    const svgContent: string = svgTemplateCompiled({ tables, connections });

    return svgContent;
};
