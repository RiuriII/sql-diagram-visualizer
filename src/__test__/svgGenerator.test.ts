import { Table } from '../interfaces';
import { svgGenerator } from '../svgGenerator';


describe('svgGenerator', () => {

	it('should generate svg', () => {
		const databaseTables: Table[] = [
			{
				tableName: 'table1',
				column: [
					{
						name: 'column1',
						type: 'INT'
					},
					{
						name: 'column2',
						type: 'VARCHAR(255)'
					},
					{
						name: 'column3',
						type: 'DATE'
					}
				],
				foreignKey: []
			}
		];

		const svg = svgGenerator(databaseTables);
		expect(svg).toMatchSnapshot();
	});

	it("Should throw an error if databaseTables is empty", () => {
		const databaseTables: Table[] = [];
		expect(() => svgGenerator(databaseTables)).toThrow('No tables found');
	})
});