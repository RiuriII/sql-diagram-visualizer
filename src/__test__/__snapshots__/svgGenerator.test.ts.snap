// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svgGenerator should generate svg 1`] = `
"
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 3000 1000" preserveAspectRatio="xMinYMin meet">
  <g id="table-0" transform="translate(100, 100)">
  
    <rect x="0" y="0" width="242" height="205" fill="#FFFFFF" stroke="black" filter="url(#shadow)"/>
    <rect x="0" y="0" width="242" height="50" fill="#8B29A6" />
    <text x="121" y="30" font-family="Arial, sans-serif" font-size="20" font-variant="small-caps" text-anchor="middle" font-weight="bold" fill="#ffffff">table1</text>
      <rect x="0" fill="#ffffff" y="50" width="241" height="40" />
      <text x="20" y="50" dy="25" font-family="Arial, sans-serif" font-size="16" fill="black" font-weight="bold">column1     <tspan dx="30" font-weight="normal">INT</tspan></text>
      <rect x="0" fill="#f5f5f5" y="90" width="241" height="40" />
      <text x="20" y="90" dy="25" font-family="Arial, sans-serif" font-size="16" fill="black" font-weight="bold">column2     <tspan dx="30" font-weight="normal">VARCHAR(255)</tspan></text>
      <rect x="0" fill="#ffffff" y="130" width="241" height="40" />
      <text x="20" y="130" dy="25" font-family="Arial, sans-serif" font-size="16" fill="black" font-weight="bold">column3     <tspan dx="30" font-weight="normal">DATE</tspan></text>
    <defs>
  <filter id="shadow" x="-10%" y="-10%" width="120%" height="120%">
    <feDropShadow dx="3" dy="3" stdDeviation="4" flood-color="#000" flood-opacity="0.2" />
  </filter>
</defs>

  </g>
</svg>
"
`;
